usdot-fhwa-stol/carma-platform
# Updated unit test to use the lanelet2_extension libraries

Updated unit test to use the lanelet2_extension libraries for the DigitalSpeedLimit regulatory element, and changed the test to use lanelet IDs instead of starting/ending points because that was not working correctly



# PR Details
## Description

The route unit test was updated to support the lanelet2_extension libraries, specifically the DigitalSpeedLimit regulatory element in osm files. 

After this was complete, the original method of route planning (grabbing a start and end point, and the code finds the closest lanelets to use) was just not working. Possibly a geo-reference location issue, bit the test would always grab the same lanelet for the starting and ending, so no routing test would occur.

This was changed so that the user puts the lanelet IDs directly into the file. There is a check that fails the unit test if either of the starting/ending lanelet IDs don't match a lanelet in the osm file.

This required a new function, routing_from_lanelets, to be created (copied from the original routing function). This then calls a pre-built function to generate a route between the two lanelets. 

## Related Issue

This change was made to support testing CAR-2671: https://usdot-carma.atlassian.net/browse/CAR-2671. Though not strictly necessary, without these updates, any files from the opendrive2lanelet converter need to have every digital_speed_limit regulatory element manually deleted to run with the original unit test. 

## Motivation and Context

This change allows osm files from the opendrive2lanelet converter to be run through the unit test, and also makes it easier to test routing on a new map. 

## How Has This Been Tested?

This was tested by testing 2 seperate files - Crossing8Course and Town02 osm files converted with the opendrive2lanelet converter and confieming that the route's shortest path was correct. This test was run in VS code in a carma docker container in a carma development VM.

    1) Libraries you should have: carma-base, carma-config, carma-msgs (integration/routing branch), 
        carma-platform (integration/routing branch), carma-utils, caram-web-ui, opendrive2lanelet.
    2) Then the test can be built with the command: 
        ./carma_build -c /workspaces/carma_ws/carma/ -a /workspaces/carma_ws/autoware.ai/ -x -m "--only-pkg-with-deps route"
    3) Update the osm file location and starting/ending IDs to match the file you want to test
    4) Run the unit test with
        catkin_make run_tests_route
    5) Confirm that the test passed and that the list of lanelet IDs does traverse from the start to the end

These changes shouldn't affect any other code, since the original functions are still there to be called. If there is a script calling this unit test, that should have the same shortcoming of having to manually edit the osm file location and starting/ending locations beforehand. 

## Types of changes


## Checklist:




[CARMA Contributing Guide](https://github.com/usdot-fhwa-stol/carma-platform/blob/develop/Contributing.md) 

I was planning to update the How to Verify HDMaps from Continental Mapping (https://vector.leidos.com/sites/STOL/STOL2/_layouts/15/WopiFrame.aspx?sourcedoc={8C47CB7D-EB7D-4AEA-9444-7587CA015A65}&file=Verify%20HD%20Maps%20from%20Continental%20Mapping.docx&action=default&DefaultItemOpen=1) documentation with the specifics of this unit test change once this pull request goes through.
